<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [6]:
import pandas as pd
import psycopg2

import warnings
warnings.filterwarnings("ignore")

In [7]:
# вставьте сюда параметры подключения из юнита 1.
# Работа с базой данных из Python


In [8]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

NameError: name 'DBNAME' is not defined

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_3_1 = f'''select /*Выбор*/
            /*Подсчитываем количество вакансий в базе*/
            count(id) as "Количество вакансий:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_3_1, connection)
df

Количество вакансий:
0                 49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_3_2 = f'''select /*Выбор*/
            /*Подсчитываем количество работодателей в базе*/
            count(id) as "Количество работодателей:"
            /*из таблицы EMPLOYERS, в которой
            хранятся данные о работодателях*/
            from public.EMPLOYERS e'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_3_2, connection)
df

Количество работодателей:
0                      23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_3_3 = f'''select /*Выбор*/
            /*Подсчитываем количество регионов в базе*/
            count(id) as "Количество регионов:" 
            /*из таблицы AREAS, в которой
            хранятся данные о регионах*/
            from public.AREAS a'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_3_3, connection)
df

Количество регионов:
0                  1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_3_4 = f'''select /*Выбор*/
            /*Подсчитываем количество сфер деятельности в базе*/
            count(id) as "Количество сфер деятельности:"  
            /*из таблицы INDUSTRIES, в которой хранятся
            данные о сферах деятельности работодателей*/
            from public.INDUSTRIES i'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_3_4, connection)
df

Количество сфер деятельности:
0                            294

***

#### ***Выводы по предварительному анализу данных:***

В представленной базе данных имеются взаимосвязаные между собой по ключу id таблицы. Это позволяет проводить анализ данных и выявлять различные характеристики взаимосвязей.

В результате проведенного предварительного анализа данных видно: 

- количество вакансий в базе данных составляет 49197 ед.;  
- количество работодателей составляет 23501 ед.;
- количество регионов составляет 1362 ед.;
- количество сфер деятельности работодателей составляет 294 ед.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_4_1 = f'''select /*Выбор*/
            /*Выбираем данные о регионах*/
            a.name as "Регион:",            
            /*Подсчитываем количество вакансий в базе*/
            count(v.id) as "Количество вакансий:"            
            /*из таблицы AREAS, в которой
            хранятся данные о регионах*/            
            from public.AREAS a            
            /*Оператор соединения таблиц; таблица  VACANCIES;
            условие: v.area_id таблицы VACANCIES
            равен a.id таблицы AREAS*/
            join public.VACANCIES v on v.area_id = a.id            
            /*Группируем по столбцу "Регион:"*/
            group by 1            
            /*Сортируем по столбцу "Количество вакансий:"
            в порядке убывания*/            
            order by 2 desc
            /*Ограничение пятью значениями*/
            limit 5'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_4_1, connection)
df

Регион:  Количество вакансий:
0           Москва                  5333
1  Санкт-Петербург                  2851
2            Минск                  2112
3      Новосибирск                  2006
4           Алматы                  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_4_2 = f'''select /*Выбор*/
            /*Подсчитываем количество вакансий в базе*/
            count(id) as "Количество вакансий:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*при условии, что заполнено хотя бы
            одно из двух полей с З/П*/
            where salary_from is not Null
            or salary_to is not Null'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_4_2, connection)
df

Количество вакансий:
0                 24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_4_3 = f'''select /*Выбор*/
            /*Вычисляем среднее значение
            для нижней границы зарплатной вилки.
            Округляем значение до целого*/
            round(AVG(v.salary_from)) as "Среднее значение З/П:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*при условии, что заполнена
            нижняя граница зарплатной вилки*/
            where v.salary_from is not Null
            /*оператор присоединения
            (уникальные значения)*/
            union            
            select /*Выбор*/
            /*Вычисляем среднее значение
            для верхней границы зарплатной вилки.
            Округляем значение до целого*/
            round(AVG(v.salary_to))
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*при условии, что заполнена
            верхняя граница зарплатной вилки*/
            where v.salary_to is not Null'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_4_3, connection)
df

Среднее значение З/П:
0                71065.0
1               110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# ТЕКСТ ЗАПРОСА:

query_4_4 = f'''select /*Выбор*/
            /*Выбираем данные "Тип рабочего графика:"*/
            schedule as "Тип рабочего графика:",
            /*Выбираем данные "Тип трудоустройства:"*/
            employment as "Тип трудоустройства:",
            /*Подсчитываем количество вакансий в базе*/
            count(id) as "Количество вакансий:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*Группируем по столбцам:
            "Тип рабочего графика:" и "Тип трудоустройства:"*/
            group by 1,2
            /*Сортируем по столбцу "Количество вакансий:"
            в порядке убывания*/
            order by 3 desc
            /*Ограничение пятью значениями*/
            limit 5'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_4_4, connection)
df

Тип рабочего графика: Тип трудоустройства:  Количество вакансий:
0           Полный день     Полная занятость                 35367
1      Удаленная работа     Полная занятость                  7802
2         Гибкий график     Полная занятость                  1593
3      Удаленная работа  Частичная занятость                  1312
4        Сменный график     Полная занятость                   940

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_4_5 = f'''select /*Выбор*/
            /*Выбираем данные "Требуемый опыт работы"*/
            experience as "Требуемый опыт работы:",            
            /*Подсчитываем количество вакансий в базе*/
            count(id) as "Количество вакансий:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*Группируем по столбцу "Требуемый опыт работы"*/
            group by 1
            /*Сортируем по столбцу "Количество вакансий:"
            в порядке возрастания*/
            order by 2 asc'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_4_5, connection)
df

Требуемый опыт работы:  Количество вакансий:
0            Более 6 лет                  1337
1              Нет опыта                  7197
2          От 3 до 6 лет                 14511
3     От 1 года до 3 лет                 26152

***

#### ***Выводы по детальному анализу вакансий:***

В результате проведенного детального анализа вакансий видно:

- максимальный спрос на трудовые ресурсы в крупных городах-миллионниках;
- средний размер зарплатной платы составляет от 70 до 110 тыс. рублей;
- максимальное количество вакансий "полная занятость - полный день" составляет 35367 ед., что в пять раз выше, чем "полная занятость - удаленная работа" - 7802 ед. Работа в офисе предпочтительнее удаленной работы.;
- максимальное количество вакансий с опытом работы "от 1 года до 3 лет" составляет 26152 ед., что в четыре раза выше, чем "без опыта работы" - 7197 ед.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# ТЕКСТ ЗАПРОСА: 

query_5_1 = f'''select /*Выбор*/
            /*Выбираем данные "Работодатель:"*/
            e.name as "Работодатель:",
            /*Подсчитываем количество вакансий в базе*/
            count(*) as "Количество вакансий:"
            /*из таблицы EMPLOYERS, в которой
            хранятся данные о работодателях*/
            from public.EMPLOYERS e
            /*Оператор соединения таблиц; таблица  VACANCIES;
            условие: v.employer_id таблицы VACANCIES
            равен e.id таблицы EMPLOYERS*/
            join VACANCIES v on v.employer_id = e.id            
            /*Группируем по столбцу "Работодатель:"*/
            group by 1
            /*Сортируем по столбцу "Количество вакансий:"
            в порядке убывания*/
            order by 2 desc
            /*Ограничение пятью значениями*/
            limit 5'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_5_1, connection)
df

Работодатель:  Количество вакансий:
0         Яндекс                  1933
1     Ростелеком                   491
2       Тинькофф                   444
3           СБЕР                   428
4  Газпром нефть                   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# ТЕКСТ ЗАПРОСА:

query_5_2 = f'''select /*Выбор*/
            /*Выбираем данные "Регион:"*/
            a.name as "Регион:",
            /*Подсчитываем количество работодателей в базе*/
            count(distinct e.id) as "Количество работодателей:",
            /*Подсчитываем количество вакансий в базе*/
            count(distinct v.id) as "Количество вакансий:"
            /*из таблицы AREAS, в которой
            хранятся данные о регионах*/
            from public.AREAS a            
            /*оператор соединения left JOIN; таблица EMPLOYERS;
            условие: e.area таблицы EMPLOYERS равен
            a.id таблицы AREAS*/            
            left join EMPLOYERS e on a.id = e.area
            /*оператор соединения left JOIN; таблица VACANCIES;
            условие: v.area_id таблицы VACANCIES равен
            a.id таблицы AREAS*/
            left join VACANCIES v on a.id = v.area_id            
            /*при условии, что не заполнен "Регион:"*/
            where v.area_id is Null            
            /*Группируем по столбцу "Регион:"*/
            group by 1
            /*Сортируем по столбцу "Количество
            работодателей:" в порядке убывания*/
            order by 2 desc
            /*Ограничение пятью значениями*/
            limit 5'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_5_2, connection)
df

Регион:  Количество работодателей:  Количество вакансий:
0              Россия                        410                     0
1           Казахстан                        207                     0
2  Московская область                         75                     0
3  Краснодарский край                         19                     0
4            Беларусь                         18                     0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# ТЕКСТ ЗАПРОСА: left join VACANCIES v on v.employer_id = e.id

query_5_3 = f'''select /*Выбор*/
            /*Выбираем данные "Работодатель:"*/
            e.name as "Работодатель:",
            /*Подсчитываем количество регионов в базе*/
            count(distinct v.area_id) as "Количество регионов:",
            /*Подсчитываем количество вакансий в базе*/
            count(distinct v.id) as "Количество вакансий:"            
            /*из таблицы EMPLOYERS, в которой
            хранятся данные о работодателях*/
            from public.EMPLOYERS e
            /*Оператор соединения таблиц; таблица  VACANCIES;
            условие: v.employer_id таблицы VACANCIES
            равен e.id таблицы EMPLOYERS*/                      
            join VACANCIES v on v.employer_id = e.id            
            /*Группируем по столбцу "Работодатель:"*/
            group by 1
            /*Сортируем по столбцу "Количество вакансий:"
            в порядке убывания*/
            order by 3 desc
            /*Ограничение пятью значениями*/
            limit 5'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_5_3, connection)
df

Работодатель:  Количество регионов:  Количество вакансий:
0         Яндекс                   181                  1933
1     Ростелеком                   152                   491
2       Тинькофф                    43                   444
3           СБЕР                    24                   428
4  Газпром нефть                    24                   331

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_5_4 = f'''select /*Выбор*/
            /*Подсчитываем количество работодателей в базе*/
            count(id) as "Количество работодателей:"
            /*из таблицы EMPLOYERS, в которой
            хранятся данные о работодателях*/
            from public.EMPLOYERS e
            /*оператор соединения left JOIN;
            таблица EMPLOYERS_INDUSTRIES; условие:
            ei.employer_id таблицы EMPLOYERS_INDUSTRIES
            равен e.id таблицы EMPLOYERS*/
            left join EMPLOYERS_INDUSTRIES ei on
            e.id = ei.employer_id
            /*при условии, что не заполнена
            "Сфера деятельности"*/
            where ei.industry_id is Null'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_5_4, connection)
df

Количество работодателей:
0                       8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [ ]:
# ТЕКСТ ЗАПРОСА: 

query_5_5 = f'''select /*Выбор*/
            /*Выбираем данные "Работодатель:"*/
            e.name as "Работодатель:"
            /*из таблицы EMPLOYERS, в которой
            хранятся данные о работодателях*/
            from public.EMPLOYERS e
            /*оператор соединения left JOIN;
            таблица EMPLOYERS_INDUSTRIES; условие:
            ei.employer_id таблицы EMPLOYERS_INDUSTRIES
            равен e.id таблицы EMPLOYERS*/
            left join EMPLOYERS_INDUSTRIES
            ei on e.id = ei.employer_id
            /*Группируем по столбцу "Работодатель:"*/
            group by 1
            /*Оператор фильтрации сгруппированных данных:
            выбираем компании, у которых указано
            четыре сферы деятельности*/
            having count(ei.industry_id) = 4
            /*Сортируем по столбцу "Работодатель:"
            в порядке возрастания*/
            order by 1 asc
            /*Ограничение значений*/
            offset 2 limit 1'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_5_5, connection)
df

Работодатель:
0          2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
# ТЕКСТ ЗАПРОСА:

query_5_6 = f'''select /*Выбор*/
            /*Подсчитываем количество работодателей в базе*/
            count(*) as "Количество работодателей:"
            /*из таблицы EMPLOYERS, в которой
            хранятся данные о работодателях*/
            from public.EMPLOYERS e            
            /*оператор соединения left JOIN;
            таблица EMPLOYERS_INDUSTRIES; условие:
            ei.employer_id таблицы EMPLOYERS_INDUSTRIES
            равен e.id таблицы EMPLOYERS*/
            left join EMPLOYERS_INDUSTRIES ei
            on e.id = ei.employer_id            
            /*оператор соединения left JOIN;
            таблица INDUSTRIES; условие:
            i.id таблицы INDUSTRIES равен
            ei.industry_id таблицы EMPLOYERS_INDUSTRIES*/            
            left join INDUSTRIES i on i.id = ei.industry_id
            /*при условии, что в сфере деятельности
            работодателя указана:
            "Разработка программного обеспечения"*/
            where i.name = 'Разработка программного обеспечения'
            '''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_5_6, connection)
df

Количество работодателей:
0                       3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
# Код для получения списка городов-милионников:

# Импортируем библиотеку requests:
import requests

# Импортируем библиотеку BeautifulSoup:
from bs4 import BeautifulSoup

# Определяем значение URL страницы для запроса:
url = "https://ru.wikipedia.org/wiki/Города-миллионеры_России"
# Делаем GET-запрос к ресурсу и результат ответа
# сохраняем в переменной response:
response = requests.get(url)

# Создаём объект BeautifulSoup, указывая html-парсер:
page = BeautifulSoup(response.text, "html.parser")
# Выводим заголовок html-страницы:
print(page.title.text, "\n")
tabl = page.find("table", class_="standard")
# Создаем путой список городов-милионников:
city = []
# В таблице выбираем тело таблицы:
tt = tabl.find("tbody")
# В теле таблицы выбираем все ряды:
qq = tt.find_all("tr")

for ss in qq:
    # В каждом ряде находим тег "а", преобразуем его в список:
    qqq = list(ss.find("a"))
    # Добавляем содержимое к списку городов:
    city.append(qqq[0])
# В первом ряду кусок от названия, отсекаем его:
city = tuple(city[1:])

# Выводим на экран список городов-милионников:
print(city)

Города-миллионеры России — Википедия 

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [ ]:
# ТЕКСТ ЗАПРОСА:

query_5_7 = f'''select /*Выбор*/
            /*Выбираем данные о регионах*/
            a.name as "Регион:",
            /*Подсчитываем количество вакансий «Яндекс»*/
            count(*) as "Количество вакансий «Яндекс»:"
            /*из таблицы EMPLOYERS, в которой
            хранятся данные о работодателях*/
            from public.EMPLOYERS e            
            /*Оператор соединения таблиц; таблица  VACANCIES;
            условие: v.employer_id таблицы VACANCIES
            равен e.id таблицы EMPLOYERS*/
            join VACANCIES v on v.employer_id = e.id
            /*Оператор соединения таблиц; таблица AREAS;
            условие: a.id таблицы AREAS
            равен v.area_id таблицы VACANCIES*/
            join AREAS a on a.id = v.area_id
            /*Условие - компания «Яндекс»,
            список регионов-миллионников*/
            where e.name = 'Яндекс' and a.name in {city}
            /*Группируем по столбцу a.id*/
            group by a.id            
            /*Оператор присоединения*/
            union all
            /*Добавяем строку Total с общим количеством
            вакансий компании 'Яндекс'*/
            select 'Total',
            sum(x.cnt)
            from
            (
            select
            count(*) as cnt
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*оператор соединения JOIN; таблица EMPLOYERS;
            условие: e.id таблицы EMPLOYERS равен
            v.employer_id таблицы VACANCIES*/
            join EMPLOYERS e on v.employer_id = e.id
            /*оператор соединения JOIN; таблица AREAS;
            условие: a.id таблицы AREAS равен
            v.area_id таблицы VACANCIES*/
            join AREAS a on a.id = v.area_id
            /*Группируем по столбцам:
            e.id, v.area_id, a.name, e.name*/
            group by e.id, v.area_id, a.name, e.name
            /*Оператор фильтрации сгруппированных данных:
            компания «Яндекс», список регионов-миллионников**/
            having e.name = 'Яндекс' and a.name in {city}
            ) as x
            /*Сортируем по столбцу "Количество вакансий:"
            в порядке возрастания*/
            order by 2 asc'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_5_7, connection)
df

Регион:  Количество вакансий «Яндекс»:
0              Омск                           21.0
1         Челябинск                           22.0
2        Красноярск                           23.0
3         Волгоград                           24.0
4    Ростов-на-Дону                           25.0
5            Казань                           25.0
6             Пермь                           25.0
7               Уфа                           26.0
8            Самара                           26.0
9         Краснодар                           30.0
10          Воронеж                           32.0
11      Новосибирск                           35.0
12  Нижний Новгород                           36.0
13     Екатеринбург                           39.0
14  Санкт-Петербург                           42.0
15           Москва                           54.0
16            Total                          485.0

***

#### ***Выводы по анализу работодателей:***

В результате проведенного детального анализа работодателей видно: 

- максимальное количество вакансий предлагает компания "Яндекс";  
- компания "Яндекс" публикует свои вакансии в 181 регионе;  
- количество вакансий компании «Яндекс» в городах-миллионниках составляет 485 ед.: Москва - 54 ед., Санкт-Петербург - 42 ед., Екатеринбург - 39 ед.;  
- в России наибольшее количество работодателей у которых нет вакансий;  
- количество работодателей, у которых не указана сфера деятельности составляет 8419 ед.  
- 3553 работодателей в качестве сферы деятельности указали "Разработка программного обеспечения";  
- компания "2ГИС" находится на третьем месте, у которой указано четыре сферы деятельности.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# ТЕКСТ ЗАПРОСА:

query_6_1 = f'''select /*Выбор*/
            /*Подсчитываем количество вакансий в базе*/
            count(id) as "Количество вакансий:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*при условии, что вакансия имеет отношение к данным,
            если в её названии содержатся слова 'data' или 'данн'*/
            where v.name ILIKE '%data%' or v.name ILIKE '%данн%'
            '''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_6_1, connection)
df

Количество вакансий:
0                  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_6_2 = f'''select /*Выбор*/
            /*Подсчитываем количество вакансий в базе*/
            count(id) as "Количество вакансий:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*Условие выбора подходящих вакансий
            для начинающего дата-сайентиста*/            
            where (v.name ILIKE '%data scientist%' 
                or v.name ILIKE '%data science%' 
                or v.name ILIKE '%исследователь данных%' 
                or v.name LIKE '%ML%' and v.name NOT LIKE '%HTML%' 
                or v.name ILIKE '%machine learning%' 
                or v.name ILIKE '%машинн%обучен%') 
                and(v.name ILIKE '%junior%'
                or lower(v.experience) = 'нет опыта'
                or lower(v.employment) = 'стажировка')
            '''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_6_2, connection)
df

Количество вакансий:
0                    51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_6_3 = f'''select /*Выбор*/
            /*Подсчитываем количество вакансий в базе*/
            count(id) as "Количество вакансий:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*Условие выбора подходящих вакансий
            для начинающего дата-сайентиста*/          
            where (v.name ILIKE '%data scientist%' 
                or v.name ILIKE '%data science%' 
                or v.name ILIKE '%исследователь данных%' 
                or v.name ILIKE '%ML%' and v.name NOT LIKE '%HTML%'
                and v.name NOT LIKE '%QML%' 
                or v.name ILIKE '%machine learning%' 
                or v.name ILIKE '%машинн%обучен%') 
                and (lower(v.key_skills) LIKE '%postgres%'
                or lower(v.key_skills) LIKE '%sql%')
            '''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_6_3, connection)
df

Количество вакансий:
0                   229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_6_4 = f'''select /*Выбор*/
            /*Подсчитываем количество вакансий в базе*/
            count(id) as "Количество вакансий:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*Условие выбора подходящих вакансий
            для начинающего дата-сайентиста*/            
            where (v.name ILIKE '%data scientist%' 
                or v.name ILIKE '%data science%' 
                or v.name ILIKE '%исследователь данных%' 
                or v.name ILIKE '%ML%' and v.name NOT LIKE '%HTML%'
                and v.name NOT LIKE '%QML%' 
                or v.name ILIKE '%machine learning%' 
                or v.name ILIKE '%машинн%обучен%') 
                and v.key_skills ILIKE '%Python%'
            '''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_6_4, connection)
df

Количество вакансий:
0                   357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_6_5 = f'''select /*Выбор*/
            /*Подсчитываем количество ключевых навыков*/            
            round(avg(length(key_skills) -
            length(replace(key_skills,CHR(9),''))+1),2)
            as "Среднее количество навыков:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v 
            /*Условие выбора подходящих вакансий
            для начинающего дата-сайентиста*/           
            where ((v.name ILIKE '%data scientist%')
                OR (lower(v.name) ILIKE '%data science%')
                OR (v.name ILIKE '%исследователь данных%')
                OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
                OR (v.name ILIKE '%machine learning%')
                OR (v.name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL))'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_6_5, connection)
df

Среднее количество навыков:
0                         6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# ТЕКСТ ЗАПРОСА:

query_6_6 = f'''select /*Выбор*/
            /*Выбираем данные "Тип требуемого опыта"*/
            v.experience as "Тип требуемого опыта:",
            /*Расчитываем среднюю З/П для дата-сайентиста*/           
            ROUND(AVG(coalesce((v.salary_from + v.salary_to) / 2,            
            v.salary_from, v.salary_to)), 0) as "Средняя З/П:"
            /*из таблицы VACANCIES, в которой
            хранятся данные о вакансиях*/
            from public.VACANCIES v
            /*Условие выбора вакансий для
            дата-сайентиста и расчета средней З/П*/
            where
            ((salary_from is not null or salary_to is not null)
                and ((v.name ILIKE '%data scientist%')
                OR (lower(v.name) ILIKE '%data science%')
                OR (v.name ILIKE '%исследователь данных%')
                OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
                OR (v.name ILIKE '%machine learning%')
                OR (v.name ILIKE '%машинн%обучен%'))
            )
            /*Группируем по столбцу "Тип требуемого опыта:"*/
            group by 1
            /*Сортируем по столбцу "Средняя З/П:"
            в порядке убывания*/
            order by 2 desc'''

In [ ]:
# РЕЗУЛЬТАТ ЗАПРОСА:

# Используем метод read_sql_query, чтобы получить
# результаты запроса сразу в виде датафрейма:
df = pd.read_sql_query(query_6_6, connection)
df

Тип требуемого опыта:  Средняя З/П:
0         От 3 до 6 лет      243115.0
1    От 1 года до 3 лет      139675.0
2             Нет опыта       74643.0

***

#### ***Выводы по предметному анализу данных:***

В результате проведенного предметного анализа данных видно: 

- количество вакансий, которые имеют отношение к данным составляет 1771 ед.; 
- количество вакансий для начинающего дата-сайентиста составляет 51 ед.;
- количество вакансий в которых в качестве ключевого навыка указан SQL или postgres составляет 229 ед.;
- количество вакансий, в которых в качестве ключевого навыка указан Python составляет 357 ед.; 
- в среднем  в вакансиях для дата-сайентиста указывают 6 ключевых навыков;
- максимальная средняя З/П для вакансий с опытом работы "От 3 до 6 лет" составляет 243,0 тыс. рублей. "От 1 года до 3 лет" составляет 140,0 тыс. рублей, "Без опыта работы" составляет 75,0 тыс. рублей.

# **Общий вывод по проекту:**

Проведенный анализ данных показал возможность с использованием данной базы данных определять количественные и качественные фарактеристики по различным типам данных, относящихся к вакансиям, работодателям, регионам и отраслям, определять различные взаимосвязи между разными разделами данных. На примере данных по вакансиям и работодателям показана возможность определения свойственных им закономерностям и тенденциям, а так же на примере вакансий относящихся к дата-сайентист, продемострирована возможность выявления данных по узким и специфическим выборкам.